docs/proposals-accepted/20240909-live-blocks-in-objectstorage.md

```diff
@@ -0,0 +1,60 @@
+
+### Problems
+
+1. How do we obtain a stable UUID to claim ownership of `thanos/sources/<uuid>/live_blocks.json`?
+
+I propose that we extend Sidecar, Ruler and Receiver to read the file `./thanos-id` on startup. This file should contain a UUID that will identify it. If this file should not exist we generate a random UUID and write this file, which should give us a reasonably stable UUID for the live time of this service.
+
+2. What happens to `thanos/sources/<uuid>/live_blocks.json` once a source churns?
+
+We will extend the compactor to delete this directory once the timestamp in `live_blocks.json` crosses a configurable threshold. Since this is asynchronous the storage gateway also needs to disregard this file if it was not updated within a configurable threshold.
+
+3. What happens if a block was deleted due to retention in Prometheus but shipper has not uploaded a new `live_blocks.json` file yet?
+
+We shall only filter blocks from the `live_blocks.json` list with a small buffer depending on the last updated timestamp. Since this list is essentially a snapshot of the blocks on disk, any block deleted because of retention will be deleted after the this timestamp. Any blocks whose range overlaps `oldest_live_block_start - (now - last_updated)` could be deleted because of retention, so they should not be pruned. Example: If we were updated 1 hour ago we should not filter the oldest block from the list. If we were updated 3 hours ago we should not filter the last 2 blocks, etc.
+
+4. What if the storage gateway has severe clockskew?
+
+Not sure how to address this yet.
+
+### Misc
+
+The `live_blocks.json` layout I propose would be:
+
+```
+[<ulid>, <ulid>, ...]
+```
+
+We should configure this behavior with a pair of flags `--shipper.upload-live-blocks` on sources `--syncer.observe-live-blocks` on storage gateway. We should also have metrics for this, I propose `thanos_shipper_last_updated_live_blocks` and `thanos_syncer_last_updated_live_blocks` with unix timestamp values and `thanos_store_live_blocks_dropped` to measure how many blocks were discarded through this.
+
+## Alternatives
+
+1. Share a Bloomfilter using Info/Series gRPC calls
+
+* We tend to do tons of `Series` calls and a bloomfilter for a decently sized bucket of 10k blocks could be enormous. Additionally the live blocks dont really change often so updating it on every `Series` call seems unnecessarily expensive.
+
+2. Shared Redis/Memcached instance
```

Comment:
Probably also worth mentioning why we don't use a filter here and opt for a JSON format with no compression - risk of false positives, UUIDs are essentially random so they don't compress well.